### PR TITLE
Correct isSubtype for intersections.

### LINKED ISSUE #2565: Crash on wildcards with enum bounds (List<Error<? extends Enum<?>>>)
Given the following Java class,
```
package cf;

import java.util.List;

public abstract class Target {

  // Broken Case:
  abstract void processErrors(List<Error<? extends Enum<?>>> errors);

  static class Error<T extends Enum<T> & Hoo> {}

  interface Hoo {}
}
```

The checker crashes with the following output:

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: Enum<? extends Enum<capture#20 of ? extends Enum<capture#20 of ? extends Enum<capture#20 of ?>>>>
  superType: Hoo
  Compilation unit: experimental/users/deltazulu/java/cf/Target.java
  Last visited tree at line 8 column 31:
    abstract void processErrors(List<Error<? extends Enum<?>>> errors);
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:25)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:154)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:338)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:28)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:67)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)
  org.checkerframework.framework.type.DefaultTypeHierarchy.castedAsSuper(DefaultTypeHierarchy.java:1085)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:465)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForTypeHierarchy.isSubtype(KeyForAnnotatedTypeFactory.java:137)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtypeOfAll(DefaultTypeHierarchy.java:289)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitIntersectionSupertype(DefaultTypeHierarchy.java:950)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Intersection(DefaultTypeHierarchy.java:538)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Intersection(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:317)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForTypeHierarchy.isSubtype(KeyForAnnotatedTypeFactory.java:137)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcardSubtype(DefaultTypeHierarchy.java:1059)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Intersection(DefaultTypeHierarchy.java:913)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Intersection(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:687)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForTypeHierarchy.isSubtype(KeyForAnnotatedTypeFactory.java:137)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:155)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2298)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkTypeArguments(BaseTypeVisitor.java:2529)
  org.checkerframework.common.basetype.BaseTypeValidator.visitParameterizedType(BaseTypeValidator.java:426)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:258)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:35)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:848)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.scan(AnnotatedTypeScanner.java:103)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:287)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:35)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:848)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.scan(AnnotatedTypeScanner.java:103)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.visit(AnnotatedTypeScanner.java:91)
  org.checkerframework.common.basetype.BaseTypeValidator.isValid(BaseTypeValidator.java:70)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateType(BaseTypeVisitor.java:3881)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateTypeOf(BaseTypeVisitor.java:3869)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitVariable(BaseTypeVisitor.java:1036)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitVariable(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:864)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:141)
  org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:683)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:800)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:81)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:133)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:384)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:339)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1008)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:515)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Initialized @NonNull Enum<? extends @Initialized @NonNull Enum<capture#20 of ? extends @Initialized @NonNull Enum<capture#20 of ? extends @Initialized @NonNull Enum<capture#20 of ?>>>>
  superType: @Initialized @NonNull Hoo
  Compilation unit: experimental/users/deltazulu/java/cf/Target.java
  Last visited tree at line 8 column 31:
    abstract void processErrors(List<Error<? extends Enum<?>>> errors);
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:25)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:154)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:338)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:28)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:67)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)
  org.checkerframework.framework.type.DefaultTypeHierarchy.castedAsSuper(DefaultTypeHierarchy.java:1085)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:465)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtypeOfAll(DefaultTypeHierarchy.java:289)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitIntersectionSupertype(DefaultTypeHierarchy.java:950)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Intersection(DefaultTypeHierarchy.java:538)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Intersection(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:317)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcardSubtype(DefaultTypeHierarchy.java:1059)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Intersection(DefaultTypeHierarchy.java:913)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Intersection(DefaultTypeHierarchy.java:47)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:687)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:179)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:155)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2298)
  org.checkerframework.checker.nullness.NullnessVisitor.commonAssignmentCheck(NullnessVisitor.java:218)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkTypeArguments(BaseTypeVisitor.java:2529)
  org.checkerframework.common.basetype.BaseTypeValidator.visitParameterizedType(BaseTypeValidator.java:426)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:258)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:35)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:848)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.scan(AnnotatedTypeScanner.java:103)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:287)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:35)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:848)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.scan(AnnotatedTypeScanner.java:103)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.visit(AnnotatedTypeScanner.java:91)
  org.checkerframework.common.basetype.BaseTypeValidator.isValid(BaseTypeValidator.java:70)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateType(BaseTypeVisitor.java:3881)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateTypeOf(BaseTypeVisitor.java:3869)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitVariable(BaseTypeVisitor.java:1036)
  org.checkerframework.checker.initialization.InitializationVisitor.visitVariable(InitializationVisitor.java:167)
  org.checkerframework.checker.initialization.InitializationVisitor.visitVariable(InitializationVisitor.java:53)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:864)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:141)
  org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:683)
  org.checkerframework.checker.initialization.InitializationVisitor.visitMethod(InitializationVisitor.java:337)
  org.checkerframework.checker.initialization.InitializationVisitor.visitMethod(InitializationVisitor.java:53)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:800)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:81)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:133)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:384)
  org.checkerframework.checker.initialization.InitializationVisitor.processClassTree(InitializationVisitor.java:295)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:339)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1008)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
2 errors
```

### PR DESCRIPTION
**Pull request by smillst**
Fixes #2565.

### CODE REVIEW

**Comment by smillst** [comment]
@mernst This build is failing because of missing documentation for a method I deleted:
>/checker-framework/framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java:279: missing documentation for areAllSubtypes

**Comment by mernst** [comment]
> @mernst This build is failing because of missing documentation for a method I deleted:
> 
> > /checker-framework/framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java:279: missing documentation for areAllSubtypes

That method isn't deleted; it is still there and not documented:  https://github.com/typetools/checker-framework/pull/2576/commits/319daef4f509ba5fff3980c1e65b4fb4bcda8c13#diff-5c4c4b5e4a6b4af9ac0c62e113ccf1a3R279
It's true that you didn't change that method, but you changed the adjacent line and that is enough to prevent the build from passing.  Can you add documentation to the method?

File: framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java [hunk 1]
```diff
@@ -949,7 +937,16 @@ protected boolean visitIntersectionSupertype(
...
         }
-        boolean result = isSubtypeOfAll(subtype, supertype.directSuperTypes());
+        boolean result = true;
+        for (AnnotatedTypeMirror aSupertype : supertype.directSuperTypes()) {
+            if (TypesUtils.isErasedSubtype(
+                            subtype.getUnderlyingType(),
+                            aSupertype.getUnderlyingType(),
+                            subtype.atypeFactory.types)
+                    && !isSubtype(subtype, aSupertype, currentTop)) {
+                result = false;
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java]: You could `break` out of the `for` loop once the result is false.

File: framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java [hunk 2]
```diff
@@ -949,7 +937,16 @@ protected boolean visitIntersectionSupertype(
         if (visitHistory.contains(subtype, supertype, currentTop)) {
             return true;
         }
-        boolean result = isSubtypeOfAll(subtype, supertype.directSuperTypes());
+        boolean result = true;
+        for (AnnotatedTypeMirror aSupertype : supertype.directSuperTypes()) {
+            if (TypesUtils.isErasedSubtype(
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java]: A comment explaining why there is this check for `isErasedSubtype` would be useful.

**Pull request merged**